### DISCUSSION TITLE
how to pass ctx nextjs to initUrqlClient next-urql

### DISCUSSION THREAD

**Discussion by edw19** [urql-graphql/urql#2019] [Q&A] [answered]
I have been trying to pass ctx to initUrqlClient but I have not succeeded, how can I pass it to it?
ese to use cookie authentication

i have looked at this example but failed

![image](https://user-images.githubusercontent.com/47750859/136624179-7ae45274-d198-480a-899d-75561186245c.png)

Can anybody  help me ?

**Comment by JoviDeCroock** [comment, marked as answer]
[Types for a client](https://formidable.com/open-source/urql/docs/api/core/#client) which is the first argument to `initUrqlClient`, I can't quite understand from this explanation what you are trying to achieve but you can leverage `fetchOptions` to set `headers`, ... but a cookie should be passed automatically if your headers are configured well as that is how cookies are supposed to work

**Reply by edw19** [reply]
what i want to achieve is this. example
I have this query that on the server is protected by the mechanism of tokens stored in cookies

![image](https://user-images.githubusercontent.com/47750859/136626057-337f8a66-d5cf-423e-81e9-9909f4887f90.png)

when I use getServerSideProps I get the unauthorized error
![image](https://user-images.githubusercontent.com/47750859/136626304-a9f955dc-2ad9-4367-a378-bc6207a291c8.png)

i am sending cookies with "include" and too in initUrqlClient
![image](https://user-images.githubusercontent.com/47750859/136626377-57e24abd-8787-404b-bf7f-331c6e835182.png)

I have noticed that the cookies object is not defined on the server

![image](https://user-images.githubusercontent.com/47750859/136626589-e03d92a1-5906-4712-a3c1-d89e855369c0.png)

**Reply by JoviDeCroock** [reply]
So `initUrqlClient` looks like this?

```js
const client = initUrqlClient({
   ...,
   fetchOptions: {
     headers: {
       cookie: ctx.req.cookie // or whatever other property it is on
     }
   }
})
```

We are talking node-node communication here so you can't rely on automatic browser forwarding like `credentials: 'include'`, within `getServerSideProps` you're in a Node process received from the initial browser call for the web-page, this then requests towards your gql server.

**Reply by edw19** [reply]
its looks like this, same that withUrqlClient

![image](https://user-images.githubusercontent.com/47750859/136629331-97ea884a-b6e6-4a79-a189-7b32e80e7a22.png)

-----
then it is not correct to make the call like that?

**Reply by JoviDeCroock** [reply]
My response shows how and more information can be found here https://github.com/node-fetch/node-fetch/issues/146 Next injects this fetch-library for Node as Node does not have fetch by default, this isn't at the end of the day an urql thing 😅 

With Next.JS it's important to keep in mind the difference when you're on the server vs on the browser, the mix makes it a bit difficult to reason about sometimes. Generally when you're doing requests like here in your props fetch you will need to include cookies in the header due to the above limitation while on the browser (withUrqlClient) credentials will be sufficient.

**Reply by edw19** [reply]
Thank you very much for the answer I thought it was something from the urql configuration now thanks to your answer it is clearer to me
